x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,276 @@
+const CHOROPLETH_POLYGONS_SOURCE_ID = 'choropleth_polygons';
+
+export function getProgressionColor(scale: number) {
+  const baseColor = euiLightVars.euiColorPrimary;
+  const adjustedScale = 0.75 * scale + 0.05; // prevents pure black & white as min/max colors.
+  if (adjustedScale < 0.5) {
+    return tint(adjustedScale * 2, baseColor);
+  }
+  if (adjustedScale > 0.5) {
+    return shade(1 - (adjustedScale - 0.5) * 2, baseColor);
+  }
+  return baseColor;
+}
+
+export function getDataRange(data: Props['data']) {
+  const firstValue = data[0] ? data[0].value : 0;
+  return data.reduce(
+    ([min, max], { value }) => [Math.min(min, value), Math.max(max, value)],
+    [firstValue, firstValue]
+  );
+}
+
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const {
+    style,
+    initialGeojsonSource,
+    geojsonKeyProperty,
+    data,
+    renderTooltip: ToolTip,
+    initialMapboxOptions
+  } = props;
+
+  const containerRef = useRef<HTMLDivElement>(null);
+  const [map, setMap] = useState<Map | null>(null);
+  const popupRef = useRef<Popup | null>(null);
+  const popupContainerRef = useRef<HTMLDivElement>(null);
+  const enableScrollZoom = useRef(false);
+  const [hoverState, setHoverState] = useState<{
+    geojsonProperties?: GeoJsonProperties;
+    data?: ChoroplethDataElement;
+  }>({});
+  // props stored in useRefs, used in the initialization side effect
+  const initialGeojsonSourceRef = useRef(initialGeojsonSource);
+  const initialMapboxOptionsRef = useRef(initialMapboxOptions);
+
+  // converts input data to a scaled value between 0 and 1
+  const getValueScale = useCallback(
+    (value: number) => {
+      const [min, max] = getDataRange(data);
+      return (value - min) / (max - min);
+    },
+    [data]
+  );
+
+  const controlScrollZoomOnWheel = useCallback((event: WheelEvent) => {
+    if (event.ctrlKey || event.metaKey) {
+      event.preventDefault();
+    } else {
+      event.stopPropagation();
+    }
+  }, []);
+
+  const updateHoverStateOnMousemove = useCallback(
+    (event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => {
+      if (
+        map &&
+        popupRef.current &&
+        data.length &&
+        map.getLayer(CHOROPLETH_LAYER_ID)
+      ) {
+        popupRef.current.setLngLat(event.lngLat);
+        const hoverFeatures = map.queryRenderedFeatures(event.point, {
+          layers: [CHOROPLETH_LAYER_ID]
+        });
+        if (hoverFeatures[0]) {
+          const geojsonProperties = hoverFeatures[0].properties;
+          // only set state when necessary since event fires so often
+          if (!isEqual(geojsonProperties, hoverState.geojsonProperties)) {
+            const matchedData = data.find(
+              ({ key }) =>
+                geojsonProperties &&
+                key === geojsonProperties[geojsonKeyProperty]
+            );
+            setHoverState({ geojsonProperties, data: matchedData });
+          }
+        } else {
+          setHoverState({});
```

Comment:
I think resetting to an object instead of `null` points to some other issues with the data structure. I think this would be better:

```js
setHoverState(null);
```